### DISCUSSION TITLE
Annotations on rules and other concepts of the Langium grammar

### DISCUSSION THREAD

**Discussion by Lotes** [eclipse-langium/langium#699] [Ideas]
Hi!

I am about to write a grammar that uses multimode lexing. We have no equivalent to pushMode/popMode like it is done in ANTLR and it also might be better to keep that actions out of the Langium grammar specification (lower complexity of Langium).

But I had a second idea that might help to write other parts like TokenBuilder, ScopeProvider and other dependents of the grammar: Annotations on rules and other concepts.

Annotations could look like this:

```
@Annotation(target='Rule', multiple=true)
interface Mode {
  name: string;
}

@Mode(name='Normal')
@Mode(name='AnotherMode')
entry TranslationUnit: stmts+=Statement+;
```

Inside of the TokenBuilder you would be able to filter the rules by mode!

One downside I see is Langium's additional complexity of handling values. You certainly need a concept for constants to not duplicate StringLiterals all over the place. And we might also need some first StandardLibrary for the @Annotation annotation...

```
//...

const ModeNormal = 'Normal';

@Mode(name=ModeNormal)
entry TranslationUnit: stmts+=Statement+;

@Mode(name=ModeNormal)
Statement: ...;
```

The big advantage is that you could annotate any constant values, and you would not have to extend Langium grammar further, because it is possible to cover additional features with a mix of annotations and smart extensions (default implementation overrides).

Another thing is the syntax of the feature. We have to think well about what we want. Entire JSONs, primitive types, forbid lambda for complexity reasons. Bli Bla Blub

WDYT?

**Comment by spoenemann** [comment]
These Java-style annotations always remind me of things that should have been part of the core language syntax (like `@Override`) or frameworks that build heavily on reflection (like Spring and many other server technologies). But that's just a syntactic view, the more important thing is what we want to do with that info.

I didn't get why you want to annotate parser rules: the lexing phase works only with the terminal rules (and keywords). It would be cool to see how a custom [TokenBuilder](https://github.com/langium/langium/blob/main/packages/langium/src/parser/token-builder.ts) using annotations of terminal rules could look like.

**Reply by Lotes** [reply]
Ahhhh! True, I mixed the concepts. I am sorry. I wanted to annotate terminal rules with their modes.

The token builder can be configured with multiple lexer modes (Chevrotain feature). So we could do something like…
```
//buildTokens method impl
const modes = getModeTokens(grammar); // : Record<string, Token[]> … get by reflecting annotations
return {modes, default: „startMode“}
```

Similar thing can be done for pushing and popping the mode.

Do not get me wrong, I do not need that syntax. I wanted to contribute the idea as inspiration for adding meta data that can be used by other parts of the language… token builder is just one example.

An alternative without grammar change could be to parse the CST comment tokens before the terminal rule… //@mode(„string“)

**Reply by spoenemann** [reply]
Yes, including that information in a comment tag could be a good solution. My concern about changing the grammar is that the additional syntax would have absolutely no effect unless you use it in your custom code.

What about defining a terminal rule syntax that includes mode changes? How much additional grammar syntax would we need?

**Reply by Lotes** [reply]
I totally understand that changing a running or established system (like Langium), can be a pain when something breaks and all dependent software projects have to migrate in order to use Langium further.

I think with annotations as explicit syntax or maybe better as comments could be a good alternative. Not only for Lexer modes. I cannot imagine other use cases right now. It is thought as a way to avoid further changes to the grammar… (I originally thought that doc-like comments would be dirtier than a concrete syntax for it).

thx for the feedback

**Reply by svallory** [reply]
Hey @spoenemann 

> My concern about changing the grammar is that the additional syntax would have absolutely no effect unless you use it in your custom code.

That's exactly my goal with #818 !

I want my grammar file (but not necessarily the grammar) to be the only source of truth for all things impacted by changes in the language. Like, TextMate Syntax grammars. Right now, the process of creating tmLanguage files is nerve-racking. Automatic language documentation generation could be another use.

But ultimately, I think a mechanism to add metadata to tokens would open up a world of possibilities for third-party tools to complement the Langium ecosystem

**Reply by msujew** [reply]
https://github.com/langium/langium/pull/866 will add JSDoc support including tags as proposed in #818.

**Reply by svallory** [reply]
Awesome, @msujew! Will the JSDoc be parsed and available programmatically? Will it be a strict JSDoc implementation or can custom annotations (e.e `@scope`) be added?

**Reply by msujew** [reply]
@svallory It's a pretty flexible implementation and doesn't really care about the actual tag values. Unless it's `@link` which has special handling. You can see the API [here](https://github.com/langium/langium/blob/main/packages/langium/src/documentation/jsdoc.ts) and some testing usage [here](https://github.com/langium/langium/blob/main/packages/langium/test/documentation/jsdoc.test.ts).

**Reply by svallory** [reply]
Looks like I'll be working on automatic TextMate syntax highlighting generation soon :)
